akkaStreamsCustomStage/src/main/scala/com/lightbend/scala/modelServer/modelServer/ModelStage.scala

```diff
@@ -1,97 +1,78 @@
-  val modelRecordIn = Inlet[ModelWithDescriptor]("modelRecordIn")
   val scoringResultOut = Outlet[Option[Double]]("scoringOut")
 
-  override val shape: ModelStageShape = new ModelStageShape(dataRecordIn, modelRecordIn, scoringResultOut)
+  override val shape: FlowShape[WineRecord, Option[Double]] = FlowShape(dataRecordIn, scoringResultOut)
 
-  class ModelLogic(shape: ModelStageShape) extends GraphStageLogicWithLogging(shape) {
+  class ModelLogic extends GraphStageLogicWithLogging(shape) {
     // state must be kept in the Logic instance, since it is created per stream materialization
     private var currentModel: Option[Model] = None
     private var newModel: Option[Model] = None
     var currentState: Option[ModelToServeStats] = None // exposed in materialized value
     private var newState: Option[ModelToServeStats] = None
 
-    override def preStart(): Unit = {
-      tryPull(modelRecordIn)
-      tryPull(dataRecordIn)
+    val setModelCB = getAsyncCallback[ModelWithDescriptor] { model =>
+      println(s"Updated model: $model")
+      newState = Some(ModelToServeStats(model.descriptor))
+      newModel = Some(model.model)
     }
 
-    setHandler(modelRecordIn, new InHandler {
-      override def onPush(): Unit = {
-        val model = grab(modelRecordIn)
-        newState = Some(new ModelToServeStats(model.descriptor))
-        newModel = Some(model.model)
-        pull(modelRecordIn)
-      }
-    })
-
     setHandler(dataRecordIn, new InHandler {
       override def onPush(): Unit = {
         val record = grab(dataRecordIn)
-        newModel match {
-          case Some(model) => {
-            // close current model first
-            currentModel match {
-              case Some(m) => m.cleanup()
-              case _ =>
-            }
-            // Update model
-            currentModel = Some(model)
-            currentState = newState
-            newModel = None
-          }
-          case _ =>
+        newModel.foreach { model =>
+          // close current model first
+          currentModel.foreach(_.cleanup())
+          // Update model
+          currentModel = Some(model)
+          currentState = newState
+          newModel = None
         }
         currentModel match {
-          case Some(model) => {
-            val start = System.currentTimeMillis()
-            val quality = model.score(record.asInstanceOf[AnyVal]).asInstanceOf[Double]
-            val duration = System.currentTimeMillis() - start
+          case Some(model) =>
+            val start = System.nanoTime()
+            val quality = model.score(record).asInstanceOf[Double]
+            val duration = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start)
             println(s"Calculated quality - $quality calculated in $duration ms")
-            currentState.get.incrementUsage(duration)
+            currentState = currentState.map(_.incrementUsage(duration))
             push(scoringResultOut, Some(quality))
```

Comment:
this push was potentially wrong previously since pull from downstream was not really used

Reply:
Yes, but you pushed without knowing if the downstream had pulled, since you did pull from preStart and ignored the onPull in the OutHandler. If downstream was slow and hadn't pulled this would have thrown exception.
The pull-push model has some nice diagrams in https://doc.akka.io/docs/akka/current/stream/stream-customize.html